### PR TITLE
Added past course run UI to dashboard

### LINKED ISSUE #939: Dashboard: Display multiple runs of the same course
This is a lower priority issue because it may be some time before we have multiple runs of courses available in the MicroMasters programs. 

The relevant cases are (see [Google Doc](https://docs.google.com/document/d/1C8JwVxLW0YZBAFwUR06754uz-uyCCEbnvt2TIMJtKk0/) for screenshots): 

### Missed Deadline to Pay

image

If the user has missed the payment deadline, they will not be able to get MicroMasters credit for the course. We show them the “re-enroll” UI, the same as if they have failed.

### Failed Course

image

If the user has failed, the course status in the right column changes to “Failed”. We then display the “Re-enroll” UI, same as if the user was auditing a course but missed the Payment Due date.

### Failed Course - Enrollment Not Open

image

If the user has failed, but the enrollment is not open, we display this status.

### Failed Course - No Future Course Runs are defined

image

This course status is a combination of “Failed” and “No Future Runs defined”

### Multiple Past Course runs

image

This mock shows the UI when a user has multiple completed past course runs.

### Re-Enrolled

image

After the student has re-enrolled, either after failing a course, or for any other reason, the current run is displayed at the top, like normal, and any “Completed Runs” are displayed below, in gray boxes.

### LINKED ISSUE #1121: (Possible bug) Unverified, non-upgradeable course runs assigned failed status
from dashboard/api.py:265

```python
    if course_enrollment.is_verified:
        if course_run.is_current:
            status = CourseRunStatus.CURRENTLY_ENROLLED
        elif course_run.is_past:
            status = CourseRunStatus.READ_CERT
        elif course_run.is_future:
            status = CourseRunStatus.WILL_ATTEND
    else:
        if (course_run.is_current or course_run.is_future) and course_run.is_upgradable:
            status = CourseRunStatus.CAN_UPGRADE
        else:
            status = CourseRunStatus.NOT_PASSED
```

If an unverified course run is current/future, but not upgradable, we are assigning a status that indicates that it is a failed course run

### LINKED ISSUE #1122: (Possible bug) Logic for fetching enroll-able CourseRuns may be incorrect
from courses/models.py:146

```python
    def get_active_enrollment_queryset():
        """
        Returns a Q object that encapsulates the logic for filtering for CourseRun objects
        that are active for enrollment.
        """
        now = datetime.now(pytz.utc)
        return (
            # there is only the start date
            (models.Q(enrollment_end=None) &
             models.Q(end_date=None) & models.Q(start_date__lte=now)) |
            # there is no enrollment date but the course has not ended yet
            (models.Q(enrollment_end=None) & models.Q(end_date__gte=now)) |
            # the enrollment end date is simply greater than now
            models.Q(enrollment_end__gte=now)
        )
```

If a CourseRun has a start_date in the future (lets say 5 days from now) and no end_date or enrollment_end, that should probably be an enroll-able course, but this logic will exclude it

### PR DESCRIPTION
**Pull request by gsidebo**
#### What are the relevant tickets?

Closes #939 
Fixes #1121 
Fixes #1122 

#### What's this PR do?

Handles courses with multiple runs per state. This is relevant in cases where the user has failed or missed an upgrade deadline for a course and we want to show them a future run, and also when the user has taken the course before multiple times.

#### Where should the reviewer start?

Probably CourseRow.js as well as dashboard/api.py where a few important changes were made

#### How should this be manually tested?

Here's a [gist](https://gist.github.com/gsidebo/478920b3ead11b4de742c076c2e8d64c) with some functions to help set up test data. Let me know if anything is confusing or doesn't work. Check #939 for the designs that these states should match

#### Screenshots (if appropriate)

ss 2016-10-13 at 21 23 02

### CODE REVIEW

**Comment by gsidebo** [comment]
FYI: I'm still working on unit tests for this. It's pretty sizable so I needed to get this up to get some eyes on it first.

**Comment by gsidebo** [comment]
unit tests fixed now after some painful merges

**Comment by gsidebo** [comment]
all unit tests fixed now. should pass

**Comment by noisecapella** [comment]
Can you look at the lint errors?

**Comment by noisecapella** [comment]
It looks like the `View on edX` link appears even if there are no course run available, can you take a look at that?

File: static/js/components/dashboard/util.js [hunk 1]
```diff
@@ -16,3 +17,9 @@ export const courseListToolTip = (text: string, id: string) => (
     </ReactTooltip>
   </div>
 );
+
+export const isCurrentlyEnrollable = (enrollmentStartDate: ?Object, now: ?moment$Moment): boolean => (
```
Comment (noisecapella) [inline static/js/components/dashboard/util.js]: Can you change `enrollmentStartDate` to also use `moment$Moment` as a type?

**Comment by noisecapella** [comment]
In the future please separate the backend and frontend work into two separate PRs, and break it up further if possible. It's difficult to do a functional review here because so many things are changing

**Comment by gsidebo** [comment]
@noisecapella the way i had this coded originally, splitting front and backend PRs would have broken the app. in hindsight i think i see an intelligent way to break them up without breaking. sorry i didnt see it sooner

**Comment by gsidebo** [comment]
@roberthouse54 is there a clear reasoning behind courses that show the 'View on edX' link vs. courses that don't? going off of this issue (#939) and these [mockups](https://projects.invisionapp.com/share/NB7VM1GSG#/screens/185611170/comments). showing/hiding the link seems pretty arbitrary. eg: the design indicates that we show it for a failed course run, and hide it for a course run that a user has missed the upgrade deadline.

**Comment by noisecapella** [comment]
If a user gets a `missed-deadline` status, wouldn't we want to show the pay button to register for the next course run if there is one? The upgrade deadline applies only to a particular course run I think

**Comment by noisecapella** [comment]
Nevermind, looks like I had the enrollment start date in the future for that run

**Comment by noisecapella** [comment]
In the mockups the 'Failed' for the previous courses is italic, can you make that change in this PR?

**Comment by noisecapella** [comment]
Actually it looks like the whole row is italic

**Comment by noisecapella** [comment]
I'm seeing some rounding weirdness:
screenshot from 2016-10-17 15-21-55

We should make a function to format grade percents consistently, here or in another PR if you prefer

**Comment by gsidebo** [comment]
@roberthouse54 see george's comment about italics above. the designs seem to have arbitrary italic font styling for some types of past course runs. can you clarify what you want here?

**Comment by noisecapella** [comment]
After a lot of testing functionality looks great! Looking at code now

**Comment by gsidebo** [comment]
@noisecapella re: missed deadline, that does bring up an interesting point. the MISSED_DEADLINE state is only assigned if your course run is current, and the code assumes that the next run to show must have a start date in the future. is it possible for these course runs to overlap? if so, it would probably make more sense to offer up a course run that is enrollable now or in the future, and (obviously) is not the same as the missed deadline run

**Comment by gsidebo** [comment]
that grade issue is strange. this is the code to render the final grade in the sub row component:

```javascript
        <Cell col={2} key="2">
          <div className="course-grade">
            { (courseRun.status === STATUS_NOT_PASSED) ? `${courseRun.final_grade}%` : ' ' }
          </div>
        </Cell>
```

it's just spitting out the value with a % at the end - no calculation involved (unless that's happening upstream). i'll look into a formatting function anyway. that's a good idea

File: dashboard/api.py [hunk 2]
```diff
@@ -207,27 +209,29 @@ def _add_run(run, mmtrack_, status):
...
-        next_run = course.get_next_run()
+        _add_run(run_status.course_run, mmtrack, CourseRunStatus.NOT_PASSED)
+        next_run = course.get_first_unexpired_run()
         if next_run is not None:
             _add_run(next_run, mmtrack, CourseStatus.OFFERED)
-        if next_run is None or run_status.course_run.pk != next_run.pk:
-            _add_run(run_status.course_run, mmtrack, CourseStatus.NOT_PASSED)
+    elif run_status.status == CourseRunStatus.MISSED_DEADLINE:
+        _add_run(run_status.course_run, mmtrack, CourseRunStatus.MISSED_DEADLINE)
+        future_run = course.get_first_future_run()
```
Comment (noisecapella) [inline dashboard/api.py]: Why are we using `get_first_future_run` instead of `get_first_unexpired_run` here?

Comment (gsidebo) [inline dashboard/api.py]: yeah, this touches on my [previous comment](https://github.com/mitodl/micromasters/pull/1389#issuecomment-254308767). the `MISSED_DEADLINE` status implies that the course run is current. `get_first_unexpired_run` would return the current course run, which we obviously don't want in this case. still, only getting future runs (ie: runs with start_date > now) is probably not correct here. one solution would be to add an optional parameter to `get_first_unexpired_run` that would exclude a specific course run/course run id. what do you think?

**Comment by roberthouse54** [comment]
For all past course runs, use italic for all text.
All courses should have the link to View on edX, as long as a link is available. The omissions on the mock were just a mistake.

File: dashboard/api.py (hunk 2 above)
Comment (gsidebo) [inline dashboard/api.py]: ok i just set this to call `get_first_unexpired_run` and tweaked that function to accept a course run to exclude

Comment (noisecapella) [inline dashboard/api.py]: Sorry for the delay, that sounds good with me

File: courses/models.py [hunk 3]
```diff
@@ -69,17 +69,20 @@ def __str__(self):
     class Meta:
         unique_together = ('program', 'position_in_program',)
 
-    def get_next_run(self):
+    def get_first_unexpired_run(self, course_run_to_exclude=None):
         """
-        Gets the next run associated with this course.
+        Gets the soonest unexpired run associated with this course.
         Note that it could be current as long as the enrollment window
         has not closed.
```
Comment (noisecapella) [inline courses/models.py]: Can you add the argument and return value to the docstring here?

File: dashboard/api.py [hunk 4]
```diff
@@ -207,27 +209,29 @@ def _add_run(run, mmtrack_, status):
...
             _add_run(next_run, mmtrack, CourseStatus.OFFERED)
     elif run_status.status == CourseRunStatus.NOT_PASSED:
-        next_run = course.get_next_run()
+        _add_run(run_status.course_run, mmtrack, CourseRunStatus.NOT_PASSED)
+        next_run = course.get_first_unexpired_run()
+        if next_run is not None:
+            _add_run(next_run, mmtrack, CourseStatus.OFFERED)
+    elif run_status.status == CourseRunStatus.MISSED_DEADLINE:
+        _add_run(run_status.course_run, mmtrack, CourseRunStatus.MISSED_DEADLINE)
+        next_run = course.get_first_unexpired_run(course_run_to_exclude=run_status.course_run)
```
Comment (noisecapella) [inline dashboard/api.py]: It seems like the exclude logic was originally applied to the `NOT_PASSED` state, should we have it there too?

File: dashboard/api.py [hunk 5]
```diff
@@ -207,27 +209,29 @@ def _add_run(run, mmtrack_, status):
...
-            _add_run(run_status.course_run, mmtrack, CourseStatus.NOT_PASSED)
     elif run_status.status == CourseRunStatus.CURRENTLY_ENROLLED:
         _add_run(run_status.course_run, mmtrack, CourseStatus.CURRENTLY_ENROLLED)
     # check if we need to check the certificate
     elif run_status.status == CourseRunStatus.CHECK_IF_PASSED:
         # if the user never passed the course she needs to enroll in the next one
         if not mmtrack.has_passed_course(run_status.course_run.edx_course_key):
-            next_run = course.get_next_run()
+            _add_run(run_status.course_run, mmtrack, CourseRunStatus.NOT_PASSED)
+            next_run = course.get_first_unexpired_run()
```
Comment (noisecapella) [inline dashboard/api.py]: Same question here about excluding `run_status.course_run`

File: static/js/components/dashboard/util.js [hunk 6]
```diff
@@ -16,3 +17,26 @@ export const courseListToolTip = (text: string, id: string) => (
     </ReactTooltip>
   </div>
 );
+
+export const isCurrentlyEnrollable = (enrollmentStartDate, now) => (
```
Comment (noisecapella) [inline static/js/components/dashboard/util.js]: Can you add flow types here?

File: static/js/components/dashboard/util.js [hunk 7]
```diff
@@ -16,3 +17,26 @@ export const courseListToolTip = (text: string, id: string) => (
...
+  enrollmentStartDate !== null &&
+    enrollmentStartDate !== undefined &&
+    enrollmentStartDate.isSameOrBefore(now || moment(), 'day')
+);
+
+export const formatGrade = (grade) => {
```
Comment (noisecapella) [inline static/js/components/dashboard/util.js]: Can you add flow types here?

File: static/js/components/dashboard/util.js [hunk 8]
```diff
@@ -16,3 +17,26 @@ export const courseListToolTip = (text: string, id: string) => (
...
+  if (_.isNil(grade) || grade === '') {
+    return '';
+  } else {
+    let gradeString;
+    if (_.isString(grade)) {
+      gradeString = _.round(Number(grade));
+    } else if (_.isInteger(grade)) {
+      gradeString = grade;
```
Comment (noisecapella) [inline static/js/components/dashboard/util.js]: I think this is unnecessary because there aren't integers in javascript, at least not in the standard types. Also, the `gradeString = _.round(Number(grade))` case should cover numbers and strings, and you've already tested against nil and an empty string, so I don't think you need the if block here at all

File: static/js/util/test_utils.js [hunk 9]
```diff
@@ -21,14 +22,31 @@ export function findCourse(courseSelector: (course: Course, program: Program) =>
...
+export function setCourseRunCount(course: Course, desiredRunCount: number, courseRunToCopy: ?CourseRun) {
+  let currentRunCount = course.runs.length;
+  if (currentRunCount < desiredRunCount) {
+    let courseRun = currentRunCount === 0 ? courseRunToCopy : course.runs[0];
+    if (!courseRun) {
+      throw new Error('Need a course run to copy.');
+    }
+    let runsNeeded = desiredRunCount - currentRunCount;
+    for (let i = 0; i < runsNeeded; i++) {
+      course.runs.push(_.cloneDeep(courseRun));
```
Comment (noisecapella) [inline static/js/util/test_utils.js]: When we're creating course runs can you modify the course run so that it's unique? We should make sure the `id` and `course_id` are unique and `position` is unique for that particular course

File: static/js/components/dashboard/util.js [hunk 10]
```diff
@@ -16,3 +17,26 @@ export const courseListToolTip = (text: string, id: string) => (
...
+    } else if (_.isNumber(grade)) {
+      // If 'grade' is a number but not an integer, round it.
+      gradeString = _.round(grade);
+    }
+    return gradeString ? `${gradeString}%` : '';
```
Comment (noisecapella) [inline static/js/components/dashboard/util.js]: This would fail if `gradeString` is `0`. It seems like `${gradeString}%` should cover all cases here since the code checks for nil above

File: static/js/components/dashboard/CourseDescription_test.js [hunk 11]
```diff
@@ -105,15 +106,14 @@ describe('CourseDescription', () => {
...
+    let firstRun = alterFirstRun(course, {course_end_date: '1999-13-92'});
+    const wrapper = shallow(<CourseDescription courseRun={firstRun} courseTitle={course.title} />);
     let elements = getElements(wrapper);
     assert.equal(elements.detailsText, '');
   });
 
   it('does not show anything when there are no runs for a course', () => {
-    let course = findCourse(course => course.runs.length === 0);
-    const wrapper = shallow(<CourseDescription course={course} />);
+    const wrapper = shallow(<CourseDescription courseRun={null} courseTitle={null} />);
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseDescription_test.js]: It looks like `CourseRow` passes an empty object to `CourseDescription` when there are no rows, not `null`, at least that's what I'm getting from `getFirstRun`. Can we make it consistent one way or the other?

File: dashboard/api.py (hunk 4 above)
Comment (gsidebo) [inline dashboard/api.py]: you can't have a `NOT_PASSED` course unless it's in the past, or 'expired' in other words. adding that kwarg would be a bit more explicit, but i'm not sure we need it. what do you think?

File: static/js/components/dashboard/util.js (hunk 6 above)
Comment (gsidebo) [inline static/js/components/dashboard/util.js]: flow was choking on some issue yesterday, and alice nor i could figure it out, but i read something last night that gave me a hint. i'll look into this again

File: static/js/components/dashboard/util.js (hunk 7 above)
Comment (gsidebo) [inline static/js/components/dashboard/util.js]: same response as above

File: static/js/components/dashboard/util.js (hunk 8 above)
Comment (gsidebo) [inline static/js/components/dashboard/util.js]: fair enough. i was aiming to avoid unnecessary operations, but the performance is quite unlikely to take a hit

File: static/js/util/test_utils.js (hunk 9 above)
Comment (gsidebo) [inline static/js/util/test_utils.js]: will do

File: static/js/components/dashboard/util.js (hunk 10 above)
Comment (gsidebo) [inline static/js/components/dashboard/util.js]: good catch

File: static/js/components/dashboard/CourseDescription.js [hunk 12]
```diff
@@ -21,12 +22,13 @@ const edxLinkBase = `${SETTINGS.edx_base_url}/courses/`;
...
-    course: Course
+    courseRun: CourseRun,
+    courseTitle: ?string
   };
 
   needsEdxLink(run: CourseRun): boolean {
-    // Any status besides 'offered' implies that the user has at some point enrolled 
-    return run.status && run.status !== STATUS_OFFERED;
+    // Any status besides 'offered' implies that the user has at some point enrolled
+    return run && run.status !== STATUS_OFFERED;
```
Comment (gsidebo) [inline static/js/components/dashboard/CourseDescription.js]: actually i just need to get rid of this function entirely. it's not needed anymore per rob's comment

File: static/js/components/dashboard/CourseRow.js [hunk 13]
```diff
@@ -1,56 +1,163 @@
...
+    let firstRun = this.getFirstRun();
+
+    let subRows = [];
+    let subRowRuns = [];
+
+    if (this.needsToEnrollAgain(firstRun)) {
+      subRowRuns.push(this.futureEnrollableRun(course));
+    }
+    let completedRuns = this.pastCourseRuns(course);
+    if (completedRuns.length > 0) {
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseRow.js]: This check is unnecessary since the concat won't do anything with an empty list

File: static/js/components/dashboard/CourseRow.js [hunk 14]
```diff
@@ -1,56 +1,163 @@
...
+      subRowRuns = subRowRuns.concat(completedRuns);
+    }
+
+    if (subRowRuns.length > 0) {
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseRow.js]: This check is also unnecessary since the for loop below won't do anything on an empty list

File: static/js/components/dashboard/CourseRow.js [hunk 15]
```diff
@@ -1,56 +1,163 @@
...
+
+    if (this.shouldDisplayGradeColumn(run)) {
+      columns.push(
+        <Cell col={2} key="2">
+          <CourseGrade courseRun={run} />
+        </Cell>
+      );
+      lastColumnSize = 4;
+    } else {
+      lastColumnSize = 6;
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseRow.js]: Can you add a test to cover this case?

**Comment by noisecapella** [comment]
Finally got through it. There's a lot of good stuff in this PR, just some minor comments here and there!

File: dashboard/api.py (hunk 4 above)
Comment (noisecapella) [inline dashboard/api.py]: Makes sense, feel free to leave it as is. Is that also true for `CHECK_IF_PASSED`?

File: static/js/components/dashboard/util.js (hunk 7 above)
Comment (noisecapella) [inline static/js/components/dashboard/util.js]: I see, feel free to leave it alone if it's being difficult

File: dashboard/api_test.py [hunk 16]
```diff
@@ -490,42 +490,44 @@ def test_info_not_passed_offered(self, mock_format):
...
     @patch('dashboard.api.format_courserun_for_dashboard', autospec=True)
     def test_info_grade(self, mock_format):
         """test for get_info_for_course for course with a course current and another not passed"""
+        self.mmtrack.configure_mock(**{'has_passed_course.return_value': False})
         with patch(
             'dashboard.api.get_status_for_courserun',
             autospec=True,
             side_effect=self.get_mock_run_status_func(
-                api.CourseRunStatus.CURRENTLY_ENROLLED, self.course_run, api.CourseRunStatus.NOT_PASSED),
+                api.CourseRunStatus.CURRENTLY_ENROLLED, self.course_run, api.CourseRunStatus.CHECK_IF_PASSED),
```
Comment (gsidebo) [inline dashboard/api_test.py]: i changed this b/c the test was failing - can't remember exactly why this fixed it. i'll look into it

**Comment by gsidebo** [comment]
@noisecapella @aliceriot i cant get flow to shut up about `formatGrade`. here it is in its current form:

```javascript
export const formatGrade = (grade: number|string|null): string => {
  if (_.isNil(grade) || grade === '') {
    return '';
  } else {
    if (_.isString(grade)) {
      grade = Number(grade);
    }
    // isFinite will return true for numbers, false for strings and NaN
    return _.isFinite(grade) ? `${_.round(grade)}%` : '';
  }
};
```

and here is flow's complaint:

```
static/js/components/dashboard/util.js:36
 36:     return _.isFinite(grade) ? `${_.round(grade)}%` : '';
                                       ^^^^^^^^^^^^^^ call of method `round`
 36:     return _.isFinite(grade) ? `${_.round(grade)}%` : '';
                                               ^^^^^ null. This type is incompatible with
325:     round(number: number, precision?: number): number;
                       ^^^^^^ number. See lib: node_modules/iflow-lodash/index.js.flow:325

static/js/components/dashboard/util.js:36
 36:     return _.isFinite(grade) ? `${_.round(grade)}%` : '';
                                       ^^^^^^^^^^^^^^ call of method `round`
 36:     return _.isFinite(grade) ? `${_.round(grade)}%` : '';
                                               ^^^^^ string. This type is incompatible with
325:     round(number: number, precision?: number): number;
                       ^^^^^^ number. See lib: node_modules/iflow-lodash/index.js.flow:325
```

i don't know if there is a way to (a) specify a type that flow will accept here, or (b) tell flow to ignore this method because it doesn't know what it's talking about. if we can't do either of those, i guess i should leave flow typing out of this file entirely? this is getting stressful

**Comment by gsidebo** [comment]
(ignore previous comment about flow typing issues 😄 )

File: dashboard/api_test.py (hunk 16 above)
Comment (gsidebo) [inline dashboard/api_test.py]: i remember now. for past course runs, we want to ignore any course runs that the user took as an audit (ie: they missed the deadline to upgrade). this if/else logic was being used to handle past course runs: `if (run_status.status == CourseRunStatus.CHECK_IF_PASSED and mmtrack.has_passed_course(run_status.course_run.edx_course_key)):`. that was made irrelevant with the choice to ignore `MISSED_DEADLINE` course runs. now essentially all we care about is non-audit course runs that have a grade, so that logic became an if/else with this condition: `run_status.status == CourseRunStatus.CHECK_IF_PASSED`. hence the change to this unit test

File: static/js/components/dashboard/CourseAction.js [hunk 17]
```diff
@@ -177,8 +175,10 @@ export default class CourseAction extends React.Component {
       }
       break;
     }
-    case STATUS_NOT_PASSED:
-      // do nothing;
+    case STATUS_MISSED_DEADLINE:
+      description = this.renderTextDescription(
+        'You missed the payment deadline and will not receive Micromasters credit for this course.'
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseAction.js]: It seems like you missed this one

File: courses/models.py [hunk 18]
```diff
@@ -69,17 +69,27 @@ def __str__(self):
...
+
+        Args:
+            course_run_to_exclude (CourseRun): A CourseRun to exclude
+                from the query
+
+        Returns (CourseRun or None): An unexpired course run
```
Comment (noisecapella) [inline courses/models.py]: The syntax here should be `Returns: CourseRun or None: An unexpired course run`, broken into separate lines if you prefer: http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html#docstrings

**Comment by noisecapella** [comment]
Can you update the MicroMasters text I mentioned in [this comment](https://github.com/mitodl/micromasters/pull/1389#discussion_r83679585)?

File: static/js/components/dashboard/CourseRow.js [hunk 19]
```diff
@@ -1,56 +1,163 @@
...
+      subRowRuns.push(this.futureEnrollableRun(course));
+    }
+    let completedRuns = this.pastCourseRuns(course);
+    if (completedRuns.length > 0) {
+      subRowRuns = subRowRuns.concat(completedRuns);
+    }
+
+    if (subRowRuns.length > 0) {
+      let i = 1;
+      for (let subRowRun of subRowRuns) {
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseRow.js]: It looks like you missed this one

**Comment by gsidebo** [comment]
fixed

File: static/js/util/test_utils.js [hunk 20]
```diff
@@ -21,14 +22,38 @@ export function findCourse(courseSelector: (course: Course, program: Program) =>
...
 
+export function setCourseRunCount(course: Course, desiredRunCount: number, courseRunToCopy: ?CourseRun) {
+  let currentRunCount = course.runs.length;
+  if (currentRunCount < desiredRunCount) {
+    let courseRun = currentRunCount === 0 ? courseRunToCopy : course.runs[0];
+    if (!courseRun) {
+      throw new Error('Need a course run to copy.');
+    }
+    let runsNeeded = desiredRunCount - currentRunCount;
+    let idMax = _.max(_.map(course.runs, run => run.id)) || 0;
```
Comment (noisecapella) [inline static/js/util/test_utils.js]: `id` needs to be unique across all programs, not just within this course

File: static/js/util/test_utils.js [hunk 21]
```diff
@@ -21,14 +22,38 @@ export function findCourse(courseSelector: (course: Course, program: Program) =>
...
+    let positionMax = _.max(_.map(course.runs, run => run.position)) || 0;
+    for (let i = 0; i < runsNeeded; i++) {
+      let newCourseRun = _.cloneDeep(courseRun);
+      positionMax++;
+      idMax++;
+      newCourseRun.position = positionMax;
+      newCourseRun.id = idMax;
+      course.runs.push(newCourseRun);
```
Comment (noisecapella) [inline static/js/util/test_utils.js]: This still needs to create a `course_id`. See the test in `constants_test.js` for the things we need to check for

**Comment by noisecapella** [comment]
Can you look at this comment? https://github.com/mitodl/micromasters/pull/1389#discussion_r83884187

File: dashboard/api.py [hunk 22]
```diff
@@ -207,27 +209,29 @@ def _add_run(run, mmtrack_, status):
...
         if next_run is not None:
             _add_run(next_run, mmtrack, CourseStatus.OFFERED)
     elif run_status.status == CourseRunStatus.NOT_PASSED:
-        next_run = course.get_next_run()
+        _add_run(run_status.course_run, mmtrack, CourseRunStatus.NOT_PASSED)
+        next_run = course.get_first_unexpired_run()
+        if next_run is not None:
+            _add_run(next_run, mmtrack, CourseStatus.OFFERED)
+    elif run_status.status == CourseRunStatus.MISSED_DEADLINE:
+        _add_run(run_status.course_run, mmtrack, CourseRunStatus.MISSED_DEADLINE)
```
Comment (noisecapella) [inline dashboard/api.py]: It looks like this case is uncovered, can you add a test for it?

screenshot from 2016-10-19 14-41-52

File: static/js/util/test_utils.js (hunk 20 above)
Comment (gsidebo) [inline static/js/util/test_utils.js]: this is creating an entirely new course object. it's not changing/altering the existing values in that giant`DASHBOARD_RESPONSE` variable, nor does it need to be concerned about the other values in there. i wrote this for the convenience of writing these tests, but it's also an intermediate step toward handling test data like we should, ie: generating the test data we need on the fly instead of depending on a giant static object that satisfies every possible condition that we want to test for. the need to check for unique id's/etc in that variable is just one symptom of the anti-pattern

Comment (noisecapella) [inline static/js/util/test_utils.js]: `course_id` and `id` need to be unique within the response data, not just within the course or program. If we were not going to check `DASHBOARD_RESPONSE` or other response objects then this function would need to create a new response object, not just a new course

Comment (gsidebo) [inline static/js/util/test_utils.js]: the point of this function is to generate data for a single Course. it's not concerned with the structure of the entire response data. i don't quite understand why this function, or the tests that use it, need to worry about that. let's chat if you want to explain further

File: static/js/util/test_utils.js [hunk 23]
```diff
@@ -21,14 +22,47 @@ export function findCourse(courseSelector: (course: Course, program: Program) =>
...
+        position: positionMax,
+        id: idMax,
+        course_id: `${newCourseRun.course_id}-new-${i}`
+      });
+      course.runs.push(newCourseRun);
+    }
+  } else if (currentRunCount > desiredRuns) {
+    course.runs = _.take(course.runs, desiredRuns);
+  }
+  course = Object.assign(course, {
```
Comment (noisecapella) [inline static/js/util/test_utils.js]: `Object.assign` assigns to the first parameter. If you wanted a shallow clone it should be something like `Object.assign({}, ...`. Can you either remove the return value or add the `{}` parameter?

Comment (gsidebo) [inline static/js/util/test_utils.js]: i don't want a shallow clone. i just want to update the object with specific key-values, like dict.update() in python. if there is a more preferred way to do this, lmk

Comment (noisecapella) [inline static/js/util/test_utils.js]: That's fine, can you remove the return value then since it's unnecessary?

File: static/js/components/dashboard/FinancialAidCard.js [hunk 24]
```diff
@@ -119,7 +119,7 @@ export default class FinancialAidCard extends React.Component {
         { courseListToolTip('filler-text', 'course-price') }
       </div>
       <div className="explanation">
-        The cost of courses in the {title} Micromasters varies
+        The cost of courses in the {title} MicroMasters varies
```
Comment (noisecapella) [inline static/js/components/dashboard/FinancialAidCard.js]: Thanks for being proactive here!
